### PR TITLE
SSL Secure Sockets [waiting for 1.13 before merging]

### PR DESCRIPTION
**Pull request by FaisalAbid**
Apologize for the formatting changes. Webstorm now uses the default dartfmt and I didn't realize till later it formatted the file.

I've highlighted the changes as comments.

Thought not fully tested, this should work. I tested it kinda hacky with localtunnel running and proxying ports, but it did work.

https://github.com/billysometimes/rethinkdb/issues/27

### CODE REVIEW

File: lib/src/net.dart [hunk 1]
```diff
@@ -1,373 +1,352 @@
...
-  int _fromBytes(List<int> data){
+  int _fromBytes(List<int> data) {
     Uint8List buf = new Uint8List.fromList(data);
     ByteData bdata = new ByteData.view(buf.buffer);
-    return bdata.getInt32(0,Endianness.LITTLE_ENDIAN);
-    }
+    return bdata.getInt32(0, Endianness.LITTLE_ENDIAN);
+  }
 
+  void _handleSocket(var socket) {
```
Comment (FaisalAbid) [inline lib/src/net.dart]: A generic method to handle  socket & securesocket connections

File: lib/src/net.dart [hunk 2]
```diff
@@ -1,373 +1,352 @@
...
-                authInfo.addAll(_toBytes(_auth_key.length));
-                authInfo.addAll(_auth_key.codeUnits);
-                _socket.add(authInfo);
-                _socket.add(_toBytes(p.VersionDummy_Protocol.JSON.value));
+  Future<Connection> reconnect([bool noreply_wait = true]) {
+    close(noreply_wait);
+
+    if (_listeners["connect"] != null) _listeners["connect"].forEach((func) => func());
+
+    if (_ssl != null && _ssl.containsKey("caCerts")) {
```
Comment (FaisalAbid) [inline lib/src/net.dart]: This is where we do the new SSL stuff

File: lib/src/net.dart [hunk 3]
```diff
@@ -1,373 +1,352 @@
...
+  final Queue<Query> _sendQueue = new Queue<Query>();
 
-    Connection(String this._db, String this._host, int this._port, String this._auth_key);
+  final Map<String, List> _listeners = new Map<String, List>();
 
-    get isClosed => _socket == null;
-    
-    use(String db){
-        _db = db;
+  Connection(String this._db, String this._host, int this._port, String this._auth_key, {Map ssl}) {
```
Comment (FaisalAbid) [inline lib/src/net.dart]: Change constructor to take optional SSL argument. Follows the same api as the official driver

**Comment by FaisalAbid** [comment]
Actually don't merge just yet, found a bug

**Comment by FaisalAbid** [comment]
Yuck, so we might have to delay this feature a bit. It works with Dart 1.13 but not Dart 1.12. The current implementation handling of SSL is shit on 1.12 since it uses one Mozilla NSS store. 1.13 has BoringSSL which works perfectly for this.

So lets keep this PR alive and I'll add 1.13 changes, then we can merge it in when 1.13 is released.

But without this, the driver works fine on the latest rethink.

**Comment by billysometimes** [comment]
I did go ahead and merge this into the 2.1 branch, we will just hold off on merging the 2.1 & master branches until the release of 1.13

**Pull request merged**